### PR TITLE
Loading indicator v2

### LINKED ISSUE #170: Show Loading indicator when processing 'highlight' actions

### PR DESCRIPTION
**Pull request by samussiah**
Closes #170

### CODE REVIEW

File: css/webcodebook.css [hunk 1]
```diff
@@ -4,11 +4,27 @@ General
...
   position: relative;
+  width:1024px;
 }
 
 .web-codebook > * {
     display: block;
-    width:1024px;
+    width: 100%;
+}
+
```
Comment (jwildfire) [inline css/webcodebook.css]: Let's add a 2nd 'background' div for the spinner with `background:white` and `opacity:0.X`

File: src/codebook/controls/filters/update.js [hunk 2]
```diff
@@ -76,17 +77,18 @@ export function update(codebook) {
...
-    //display the loading indicator
-    codebook.loadingIndicator.style('display', 'block');
-    //wait by the quarter second until the loading indicator is visible to re-render everything
-    const loading = setInterval(() => {
-      const display = codebook.loadingIndicator.style('display');
-      if (display === 'block') {
+  var filters = codebook.controls.wrap
+    .selectAll('.filterCustom select')
+    .on('change', function(d) {
+      indicateLoading(codebook, '#loading-indicator', () => {
```
Comment (jwildfire) [inline src/codebook/controls/filters/update.js]: This seems tautological (but it does seem to work). Doesn't this mean to show the `#loading-indicator` spinner until the `#loading-indicator` spinner appears?!

Comment (jwildfire) [inline src/codebook/controls/filters/update.js]: Same as comment above.

File: src/codebook/controls/groups/update.js [hunk 3]
```diff
@@ -39,25 +40,14 @@ export function update(codebook) {
   );
   groupOptions.sort((a, b) => columns.indexOf(a) - columns.indexOf(b));
   groupSelect.on('change', function() {
-    //display loading indicator
-    codebook.loadingIndicator.style('display', 'block');
+    indicateLoading(codebook, '#loading-indicator', () => {
+      if (this.value !== 'None') codebook.config.group = this.value;
```
Comment (jwildfire) [inline src/codebook/controls/groups/update.js]: Same comment as above.

File: src/codebook/util/indicateLoading.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import { select as d3select } from 'd3';
+
+export default function indicateLoading(codebook, element, callback) {
+  codebook.loadingIndicator.style('display', 'block');
+  //wait by the centisecond until the loading indicator is visible
+  const loading = setInterval(() => {
+    const laidOut = d3select(element).property('offsetwidth') > 0;
+    if (!laidOut) {
```
Comment (jwildfire) [inline src/codebook/util/indicateLoading.js]: Check that the target element does not have "display:none" before attempting to render?

File: src/codebook/controls/filters/update.js (hunk 2 above)
Comment (samussiah) [inline src/codebook/controls/filters/update.js]: It's definitely recursive to some extent.  In effect the call starts the spinner then summaryTable.draw() and dataListing.init() pick up where it leaves off.

**Comment by samussiah** [comment]
Updated @jwildfire

File: src/codebook/controls/filters/update.js (hunk 2 above)
Comment (jwildfire) [inline src/codebook/controls/filters/update.js]: Ahhhhhhh, I get it. Basically several different spinners are initialized by these calls. Works for me. 👍

**Pull request merged**